### DISCUSSION TITLE
Type casting without instances

### DISCUSSION THREAD

**Discussion by ChristopherHaas** [dotnet/csharplang#7889] [Language Ideas]
Suppose you have an interface with an abstract static method. Ex:

```
public interface ITest<T> where  T : ITest<T>
{
abstract static T Create(…) 
}

```

Then, inside another method, You may want to call this factory method, if possible. Something like:

```
public void Create<T>() 
{
   if (T is ITest<T> otherT) {
      return otherT.Create();
   }
else
  {
// Reflection fallback, fail, or something else
  }
}
```

However, it does not seem possible to describe this kind of runtime type testing in C# (without reflection, IL emit etc.). It would be beneficial for the overall _abstract static in interfaces_ feature to allow some kind of typecasting without instances.

Am I missing something, or am I correct in concluding that this cannot currently be done?

**Comment by HaloFour** [comment]
There is currently no way in the language or runtime to conditionally constraint a generic type parameter within the body of a method.   The closest you get today is via reflection.

See: #6308
